README.md

# Animated Sidebar

[![pub package](https://img.shields.io/pub/v/animated_sidebar.svg?style=flat-square)](https://pub.dev/packages/animated_sidebar)
[![Libraries.io for GitHub](https://img.shields.io/librariesio/github/kesimo/animated_sidebar.svg?style=flat-square)](https://libraries.io/github/kesimo/animated_sidebar) 
[![License](https://img.shields.io/badge/License-MIT%20-green.svg?style=flat-square)](https://opensource.org/licenses/MIT)

A highly customizable and styleable collapsable sidebar plugin for Flutter, optimized for web and desktop applications.

## Example

![example-ui](https://github.com/kesimo/animated_sidebar/blob/main/media/example.gif)

## Installation

Add the following to your pubspec.yaml file:

```bash  
dependencies:  
 animated_sidebar: ^1.0.0   
```  

or use the following command:

```bash  
flutter pub add animated_sidebar  
```  

Add the following import to your dart file:

```dart  
import 'package:animated_sidebar/animated_sidebar.dart';  
```  

## Usage

### Sidebar Items
define a list of `SidebarItem` objects:

```dart  
import 'package:animated_sidebar/animated_sidebar.dart';  
  
final List<SidebarItem> items = [  
  SidebarItem(icon: Icons.home, text: 'Home'),  
  SidebarItem(icon: Icons.notifications, text: 'Notifications'),  
  SidebarItem(icon: Icons.person, text: 'Management'),  
];  
```

### Child Items
it is also possible to define multiple `SidebarChildItem` for every `SidebarItem`.

```dart  
import 'package:animated_sidebar/animated_sidebar.dart';  
  
final List<SidebarItem> items = [  
  SidebarItem(icon: Icons.home, text: 'Home'),  
  SidebarItem(  
    icon: Icons.person,  
    text: 'Management',  
    children: [  
	  SidebarChildItem(icon: Icons.person, text: 'Users'),  
	  SidebarChildItem(icon: Icons.verified_user, text: 'Roles'),  
    ],  
  ),
];  
```
The Item Containing the Children is Collapsed and expand only on tap. If the Current selected item is a `SidebarChildItem` the overlying Item keeps expanded


### Default usage of `AnimatedSidebar`

```dart
import 'package:animated_sidebar/animated_sidebar.dart';

@override
Widget build(BuildContext context) {
  return Scaffold(
    backgroundColor: Theme.of(context).scaffoldBackgroundColor,
    body: Row(
      children: [
        AnimatedSidebar(
          expanded: MediaQuery.of(context).size.width > 600,
          items: items,
          selectedIndex: 0,
          onItemSelected: (index) => print(index),
          headerIcon: Icons.ac_unit_sharp,
          headerIconColor: Colors.amberAccent,
          headerText: 'Example',
        ),
        Center(
          child: Text(
            'Page: $activeTab',
            style: Theme.of(context).textTheme.headline3,
          ),
        ),
      ],
    ),
  );
}
```


### Use the `AnimatedSidebar` widget and *handle state external* :

```dart
import 'package:animated_sidebar/animated_sidebar.dart';

int activeTab = 0;

@override
Widget build(BuildContext context) {
  return Scaffold(
    backgroundColor: Theme.of(context).scaffoldBackgroundColor,
    body: Row(
      children: [
        AnimatedSidebar(
          expanded: MediaQuery.of(context).size.width > 600,
          items: items,
          selectedIndex: activeTab,
          autoSelectedIndex: false, // must be false if you want to handle state external
          onItemSelected: (index) =>
              setState(() => activeTab = index),
          headerIcon: Icons.ac_unit_sharp,
          headerIconColor: Colors.amberAccent,
          headerText: 'Example',
        ),
        Center(
          child: Text(
            'Page: $activeTab',
            style: Theme.of(context).textTheme.headline3,
          ),
        ),
      ],
    ),
  );
}
```

CHANGELOG.md
## 1.0.1

* Code formatting and cleanup

## 1.0.0

* Add `SidebarChildItem` to create a collapsible sidebar item
* Update `README.md`

## 0.2.0

* Add `autoSelectedIndex` parameter to `AnimatedSidebar` constructor. To handle state internal by default 
* Updated `README.md`

## 0.1.0

* Added documentation for AnimatedSidebar and SidebarItem
* Removed `footer` property from AnimatedSidebar
* Removed `frameMargin` property from AnimatedSidebar

## 0.0.1

* Initial release.


test/animated_sidebar_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:animated_sidebar/animated_sidebar.dart';

void main() {
  // predefine items
  List<SidebarItem> items() => [
        SidebarItem(
          icon: Icons.home,
          text: 'Home',
        ),
        SidebarItem(
          icon: Icons.notification_important_rounded,
          text: 'Notifications', //global variables, properties, etc.
        ),
        SidebarItem(
          icon: Icons.person,
          text: 'Management', //api, tokens, users, access rules etc.
        ),
        SidebarItem(
          icon: Icons.abc,
          text: 'Integrations', //api, tokens, users, access rules etc.
        ),
        SidebarItem(
          icon: Icons.settings,
          text: 'Settings', //api, tokens, users, access rules etc.
        ),
      ];

  // test sidebar widget
  testWidgets('Sidebar widget test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(
      MaterialApp(
        builder: (context, child) {
          return Scaffold(
            body: Container(
              margin: const EdgeInsets.fromLTRB(16, 24, 0, 24),
              child: AnimatedSidebar(
                expanded: true,
                items: items(),
                selectedIndex: 0,
                onItemSelected: (index) => {},
                duration: const Duration(milliseconds: 250),
                minSize: 90,
                maxSize: 250,
                itemIconSize: 26,
                itemIconColor: Colors.white,
                itemHoverColor: Colors.grey.withOpacity(0.3),
                itemSelectedColor: Colors.grey,
                itemSelectedBorder: const BorderRadius.all(
                  Radius.circular(5),
                ),
                itemMargin: 16,
                itemSpaceBetween: 10,
                headerIcon: Icons.menu,
                headerIconSize: 30,
                headerIconColor: Colors.deepPurple,
                headerTextStyle: const TextStyle(
                    fontSize: 22,
                    fontWeight: FontWeight.w600,
                    color: Colors.white),
                headerText: ' TimeTasks',
              ),
            ),
          );
        },
      ),
    );

    // Verify that our counter starts at 0.
    expect(find.text('Home'), findsOneWidget);
    expect(find.text('Management'), findsOneWidget);
    expect(find.text('Notifications'), findsOneWidget);
    expect(find.text('Integrations'), findsOneWidget);
    expect(find.text('Settings'), findsOneWidget);
  });
}


lib/src/sidebar.dart
import 'package:animated_sidebar/src/sidebar_item.dart';
import 'package:flutter/material.dart';

/// AnimatedSidebar is a stateful widget that displays a sidebar that can
/// Collapse and Expand.
///
/// see: [git](https://github.com/kesimo/animated_sidebar)
class AnimatedSidebar extends StatefulWidget {
  /// A list of SidebarItem objects. Contains the text and icon for each item.
  final List<SidebarItem> items;

  /// This function is called when a sidebar item is tapped.
  /// It takes the index of the tapped item in [items] as a parameter.
  final void Function(int index) onItemSelected;

  /// The minimum width of the [AnimatedSidebar] when it is collapsed.
  final double minSize;

  /// The maximum width of the [AnimatedSidebar] when it is expanded.
  final double maxSize;

  /// The index of the selected item. based on the index of the element in the [items] list
  /// this is used to highlight the selected item.
  ///
  /// on page change, this value should be updated.
  ///
  /// NOTE:
  ///  * Prefer using [autoSelectedIndex] instead of this. so the selected item changes itself.
  ///    and you don't have to update it manually what causes a rebuild.
  ///  * If [autoSelectedIndex] is true, this value will be used as Initial value.
  final int selectedIndex;

  /// If true, the selected item will be automatically updated based on the last
  /// item that was tapped.
  ///
  /// NOTE:
  /// * If [autoSelectedIndex] is true, [selectedIndex] will be used as Initial value.
  final bool autoSelectedIndex;

  /// Used to determine the initial state of the sidebar.
  final bool expanded;

  /// The margin of the sidebar.
  final EdgeInsets margin;

  /// The duration of the animation.
  /// The default value is 250 milliseconds.
  final Duration duration;

  /// Used to determine the animation curve.
  /// The default value is [Curves.easeInOut].
  final Curve curve;

  /// The size of the [SidebarItem.icon] for all [items] in the sidebar.
  /// The default value is 32.0.
  final double itemIconSize;

  /// The color of the [SidebarItem.icon] for all [items] in the sidebar.
  /// The default value is [Colors.white].
  final Color itemIconColor;

  /// Set the text style of the [SidebarItem.text].
  final TextStyle itemTextStyle;

  /// The space between each item in [items].
  /// The default value is 8.0.
  final double itemSpaceBetween;

  /// The Color of the [SidebarItem] when it is selected
  /// OR: the [selectedIndex] is equal to the index of the item.
  ///
  /// The default value is [Colors.indigoAccent].
  final Color? itemSelectedColor;

  /// The Color of the [SidebarItem] when it is hovered.
  ///
  /// The default value is [Colors.indigoAccent] with 0.3 opacity.
  final Color? itemHoverColor;

  /// The border radius of the overlay on the selected [SidebarItem].
  final BorderRadiusGeometry itemSelectedBorder;

  /// The margin of the [SidebarItem].
  final double itemMargin;

  /// The [switchIconExpanded] that is displayed on the bottom of the sidebar
  /// when the sidebar is expanded.
  /// On tap, the sidebar will be collapsed.
  final IconData? switchIconExpanded;

  /// The [switchIconCollapsed] that is displayed on the bottom of the sidebar
  /// when the sidebar is collapsed.
  /// On tap, the sidebar will be expanded.
  final IconData? switchIconCollapsed;

  /// The decoration of the whole [AnimatedSidebar] frame.
  /// The default value is a [BoxDecoration] with a [BoxShadow] and a [BorderRadius] of 10.
  final BoxDecoration frameDecoration;

  /// The [header] can be used instead of combination of [headerIcon] and [headerText].
  /// for a custom header widget.
  ///
  /// Note:
  ///
  ///   * if [header] is not null, [headerIcon] and [headerText] will be ignored.
  ///   * the [header] should be a responsive widget.
  final Widget? header;

  /// The [headerIcon] is displayed on the top of the sidebar.
  ///
  /// if null, only the [headerText] will be displayed.
  final IconData? headerIcon;

  /// the size of the [headerIcon].
  final double? headerIconSize;

  /// The color of the [headerIcon].
  final Color? headerIconColor;

  /// the style of the [headerText].
  ///
  /// Overflow will be handled by [TextOverflow.fade].
  final TextStyle? headerTextStyle;

  /// The [headerText] is displayed on the top of the sidebar.
  final String? headerText;

  const AnimatedSidebar({
    Key? key,
    required this.items,
    this.selectedIndex = 0,
    this.autoSelectedIndex = true,
    required this.onItemSelected,
    this.expanded = true,
    this.margin = const EdgeInsets.all(16),
    this.minSize = 90,
    this.maxSize = 250,
    this.duration = const Duration(milliseconds: 250),
    this.curve = Curves.easeInOut,
    this.itemIconSize = 32,
    this.itemIconColor = Colors.white,
    this.itemTextStyle = const TextStyle(
        fontSize: 20, fontWeight: FontWeight.w500, color: Colors.white),
    this.itemSpaceBetween = 8,
    this.itemSelectedColor =
        const Color.fromRGBO(48, 79, 254, 1), //MaterialIndigo
    this.itemHoverColor =
        const Color.fromRGBO(48, 79, 254, 0.3), //MaterialIndigo
    this.itemSelectedBorder = const BorderRadius.all(Radius.circular(5)),
    this.itemMargin = 16,
    this.switchIconExpanded = Icons.arrow_back_rounded,
    this.switchIconCollapsed = Icons.arrow_forward_rounded,
    this.frameDecoration = const BoxDecoration(
      color: Colors.black,
      borderRadius: BorderRadius.all(Radius.circular(10)),
      boxShadow: [
        BoxShadow(
          color: Color.fromRGBO(66, 66, 66, 0.75),
          spreadRadius: 0,
          blurRadius: 20,
          offset: Offset(0, 10),
        ),
      ],
    ),
    this.header,
    this.headerIcon,
    this.headerIconSize = 32,
    this.headerIconColor = Colors.blueAccent,
    this.headerTextStyle = const TextStyle(
        fontSize: 22, fontWeight: FontWeight.w500, color: Colors.white),
    this.headerText,
  })  : assert((headerIcon != null && headerText != null) ^ (header != null)),
        super(key: key);

  @override
  State<AnimatedSidebar> createState() => _AnimatedSidebarState();
}

class _AnimatedSidebarState extends State<AnimatedSidebar>
    with TickerProviderStateMixin {
  bool _inAnimation = false;
  bool _expanded = true;
  int _onHoverIndex = -1;
  int _selectedIndex = 0;
  int _expanedIndex = -1;

  void _resize() {
    setState(() {
      _expanded = !_expanded;
      _inAnimation = true;
    });
  }

  void _setInAnimation(bool value) {
    setState(() {
      _inAnimation = value;
    });
  }

  void _setOnHover(int index) {
    setState(() {
      _onHoverIndex = index;
    });
  }

  void _setExitHover() {
    setState(() {
      _onHoverIndex = -1;
    });
  }

  double _calculateItemOffset() {
    return (widget.minSize - widget.itemMargin * 2 - widget.itemIconSize) / 2;
  }

  double _calculateHeaderItemOffset() {
    return (widget.minSize -
            widget.itemMargin * 2 -
            (widget.headerIconSize ?? 0)) /
        2;
  }

  void _setSelectedIndex(int index) {
    if (widget.autoSelectedIndex) {
      setState(() {
        _selectedIndex = index;
      });
    }
  }

  @override
  void initState() {
    _expanded = widget.expanded;
    _selectedIndex = widget.selectedIndex;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      margin: widget.margin,
      width: _expanded ? widget.maxSize : widget.minSize,
      duration: widget.duration,
      curve: widget.curve,
      onEnd: () => _setInAnimation(false),
      child: _buildFrame(),
    );
  }

  Widget _buildFrame() {
    return Container(
      height: double.infinity,
      decoration: widget.frameDecoration,
      child: _buildChild(),
    );
  }

  Widget _buildChild() {
    return Container(
      margin: EdgeInsets.all(widget.itemMargin),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        children: [
          widget.header != null ? _buildCustomHeader() : _buildIconTextHeader(),
          Container(
            margin: const EdgeInsets.only(top: 16, bottom: 24),
            height: 1,
            color: Colors.grey,
          ),
          Expanded(
            child: LayoutBuilder(
                builder: (BuildContext context, BoxConstraints constraints) {
              return ConstrainedBox(
                constraints: constraints,
                child: SingleChildScrollView(
                  child: Column(
                    children: _buildMenuItems(),
                  ),
                ),
              );
            }),
          ),
          _buildSwitchButton(),
        ],
      ),
    );
  }

  IconButton _buildSwitchButton() {
    return IconButton(
        onPressed: () {
          setState(() {
            if (_inAnimation) return;
            _resize();
          });
        },
        icon: Icon(
            _expanded ? widget.switchIconExpanded : widget.switchIconCollapsed,
            color: _inAnimation ? Colors.transparent : widget.itemIconColor));
  }

  Widget _buildIconTextHeader() {
    return Container(
      padding: EdgeInsets.symmetric(horizontal: _calculateHeaderItemOffset()),
      height: 64,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        children: [
          Icon(
            widget.headerIcon,
            color: widget.headerIconColor,
            size: widget.headerIconSize,
          ),
          _expanded || _inAnimation
              ? Flexible(
                  child: Text(
                    widget.headerText ?? 'missing',
                    overflow: TextOverflow.fade,
                    maxLines: 1,
                    softWrap: false,
                    style: widget.headerTextStyle,
                  ),
                )
              : const SizedBox.shrink(),
        ],
      ),
    );
  }

  Widget _buildCustomHeader() {
    return widget.header ?? Container();
  }

  List<Widget> _buildMenuItems() {
    List<Widget> items = [];
    int index = 0;
    for (int i = 0; i < widget.items.length; i++) {
      if (widget.items[i].children.isNotEmpty) {
        items.add(
          _buildMultiItem(index, widget.items[i]),
        );
        index += widget.items[i].children.length + 1;
      } else {
        items.add(
          _buildSingleItem(index, widget.items[i], false, false),
        );
        index++;
      }
      items.add(
        SizedBox(
          height: widget.itemSpaceBetween,
        ),
      );
    }
    return items;
  }

  Widget _buildSingleItem(
      int index, SidebarItem item, bool isChild, bool isHeader) {
    Color? color = index ==
            (widget.autoSelectedIndex ? _selectedIndex : widget.selectedIndex)
        ? widget.itemSelectedColor
        : (index == _onHoverIndex ? widget.itemHoverColor : Colors.transparent);

    bool colorHeader = !_expanded &&
        !_inAnimation &&
        isHeader &&
        index >= _expanedIndex &&
        index <
            _expanedIndex +
                widget.items[index - _expanedIndex].children.length +
                1;

    return MouseRegion(
      cursor: SystemMouseCursors.click,
      onHover: (_) => _setOnHover(index),
      onExit: (_) => _setExitHover(),
      child: GestureDetector(
        child: Container(
          width: double.infinity,
          padding: const EdgeInsets.symmetric(
            vertical: 4,
          ),
          decoration: BoxDecoration(
              borderRadius: widget.itemSelectedBorder,
              color: colorHeader ? widget.itemSelectedColor : color),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.start,
            mainAxisSize: MainAxisSize.min,
            children: [
              SizedBox(
                width: _calculateItemOffset(),
              ),
              Icon(
                item.icon,
                color: widget.itemIconColor,
                size: widget.itemIconSize,
              ),
              _expanded || _inAnimation
                  ? const Padding(padding: EdgeInsets.only(left: 8))
                  : const SizedBox.shrink(),
              _expanded || _inAnimation
                  ? Flexible(
                      child: Text(
                        item.text,
                        overflow: TextOverflow.fade,
                        maxLines: 1,
                        softWrap: false,
                        style: isChild
                            ? widget.itemTextStyle.copyWith(
                                color: widget.itemTextStyle.color
                                    ?.withOpacity(0.6))
                            : widget.itemTextStyle,
                      ),
                    )
                  : const SizedBox.shrink(),
              _expanded && !_inAnimation && isHeader
                  ? Padding(
                      padding: const EdgeInsets.fromLTRB(5, 0, 0, 0),
                      child: Icon(
                        Icons.keyboard_arrow_down,
                        size: widget.itemIconSize,
                        color: widget.itemIconColor,
                      ),
                    )
                  : const SizedBox.shrink(),
            ],
          ),
        ),
        onTap: () {
          if (isHeader) {
            setState(() {
              _expanedIndex = index;
            });
          } else if (!isChild) {
            setState(() {
              _expanedIndex = -1;
            });
          }
          _setSelectedIndex(index);
          widget.onItemSelected(index);
        },
      ),
    );
  }

  Widget _buildMultiItem(int baseIndex, SidebarItem item) {
    List<Widget> childItems = [];
    for (int i = 0; i < item.children.length; i++) {
      childItems.add(
        SizedBox(
          height: widget.itemSpaceBetween,
        ),
      );
      childItems.add(
        _buildSingleItem(
            baseIndex + i + 1,
            SidebarItem(
              text: item.children[i].text,
              icon: item.children[i].icon,
            ),
            true,
            false),
      );
    }
    return Column(
      children: [
        _buildSingleItem(baseIndex, item, false, true),
        Container(
          margin: const EdgeInsets.only(left: 12),
          child: _expanded &&
                  !_inAnimation &&
                  ((_selectedIndex >= baseIndex &&
                          _selectedIndex <= baseIndex + childItems.length) ||
                      _expanedIndex == baseIndex)
              ? Column(
                  children: childItems,
                )
              : const SizedBox.shrink(),
        ),
      ],
    );
  }
}


lib/src/sidebar_item.dart
import 'package:flutter/material.dart';

/// A [SidebarItem] is a Object that contains the information of a sidebar item.
/// It contains the text and the icon of the item.
/// Possible to add child items to the item and make it expanding.
/// It is used in the [AnimatedSidebar] widget. (See [AnimatedSidebar.items])
class SidebarItem {
  SidebarItem({
    required this.text,
    required this.icon,
    this.children = const [],
  });

  /// The text of the item.
  final String text;

  /// The icon of the item.
  final IconData icon;

  /// The optional child items of the item.
  final List<SidebarChildItem> children;
}

/// A [SidebarChildItem] is a Object that contains the information of a sidebar item.
/// It contains the text and the icon of the item and should only be used as a child of a [SidebarItem].
/// It is used in the [AnimatedSidebar] widget. (See [AnimatedSidebar.items])
class SidebarChildItem {
  SidebarChildItem({
    required this.text,
    required this.icon,
  });

  /// The text of the item.
  final String text;

  /// The icon of the item.
  final IconData icon;
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:animated_sidebar/animated_sidebar.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int activeTab = 0;

  final List<SidebarItem> items = [
    SidebarItem(
      icon: Icons.home,
      text: 'Home',
    ),
    SidebarItem(
      icon: Icons.notifications,
      text: 'Notifications',
    ),
    SidebarItem(
      icon: Icons.person,
      text: 'Management',
      children: [
        SidebarChildItem(
          icon: Icons.person,
          text: 'Users',
        ),
        SidebarChildItem(
          icon: Icons.verified_user,
          text: 'Roles',
        ),
        SidebarChildItem(
          icon: Icons.key,
          text: 'Permissions',
        ),
      ],
    ),
    SidebarItem(
      icon: Icons.abc,
      text: 'Integrations',
    ),
    SidebarItem(
      icon: Icons.settings,
      text: 'Settings',
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Theme.of(context).scaffoldBackgroundColor,
      body: SafeArea(
        child: Row(
          children: [
            AnimatedSidebar(
              margin: const EdgeInsets.fromLTRB(16, 24, 0, 24),
              expanded: MediaQuery.of(context).size.width > 600,
              items: items,
              // use this to set the active tab if you want to control it from outside
              // combined with autoSelectedIndex set to false
              // if you don't set autoSelectedIndex to false, the widget will
              // automatically set the active tab and selected item is used only
              // to set the initial value
              selectedIndex: activeTab,
              autoSelectedIndex: false,
              onItemSelected: (index) => setState(() => activeTab = index),
              duration: const Duration(milliseconds: 1000),
              frameDecoration: const BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  tileMode: TileMode.repeated,
                  colors: [
                    Color.fromRGBO(66, 66, 74, 1),
                    Color.fromRGBO(25, 25, 25, 1),
                  ],
                ),
                borderRadius: BorderRadius.all(Radius.circular(10)),
                boxShadow: [
                  BoxShadow(
                    color: Color.fromRGBO(66, 66, 66, 0.75),
                    spreadRadius: 0,
                    blurRadius: 20,
                    offset: Offset(0, 10),
                  ),
                ],
              ),
              minSize: 90,
              maxSize: 250,
              itemIconSize: 26,
              itemIconColor: Colors.white,
              itemHoverColor: Colors.grey.withOpacity(0.3),
              itemSelectedColor: Colors.grey.withOpacity(0.3),
              itemTextStyle: const TextStyle(color: Colors.white, fontSize: 20),
              itemSelectedBorder: const BorderRadius.all(
                Radius.circular(5),
              ),
              itemMargin: 16,
              itemSpaceBetween: 10,
              headerIcon: Icons.ac_unit_sharp,
              headerIconSize: 30,
              headerIconColor: Colors.amberAccent,
              headerTextStyle: const TextStyle(
                  fontSize: 24,
                  fontWeight: FontWeight.w600,
                  color: Colors.white),
              headerText: ' Example',
            ),
            Expanded(
              child: _buildPage(activeTab),
            ),
          ],
        ),
      ),
    );
  }

  //minimal version
  // @override
  // Widget build(BuildContext context) {
  //   return Scaffold(
  //     backgroundColor: Theme.of(context).scaffoldBackgroundColor,
  //     body: Row(
  //       children: [
  //         AnimatedSidebar(
  //           expanded: MediaQuery.of(context).size.width > 600,
  //           items: items,
  //           selectedIndex: activeTab,
  //           onItemSelected: (index) =>
  //               setState(() => activeTab = index),
  //           headerIcon: Icons.ac_unit_sharp,
  //           headerIconColor: Colors.amberAccent,
  //           headerText: 'Example',
  //         ),
  //         Center(
  //           child: Text(
  //             'Page: $activeTab',
  //             style: Theme.of(context).textTheme.headline3,
  //           ),
  //         ),
  //       ],
  //     ),
  //   );
  // }

  Widget _buildPage(int idx) {
    // return a page full of skeleton cards
    return Wrap(
      children: List.generate(8, (index) {
        double width = idx % 2 == 0 ? double.infinity : 500;
        int count = Random().nextInt(2) + 3;
        return Container(
            margin: const EdgeInsets.fromLTRB(24, 24, 24, 0),
            width: width,
            height: 250,
            decoration: BoxDecoration(
              color: Colors.grey.withOpacity(0.3),
              borderRadius: BorderRadius.circular(10),
            ),
            // create a skeleton card
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                for (int i = 0; i < count; i++)
                  Container(
                    width: Random().nextInt(100).toDouble() + 200,
                    height: i == 0 ? 30 : 20,
                    margin: const EdgeInsets.all(16),
                    decoration: BoxDecoration(
                      color: Colors.grey.withOpacity(0.3),
                      borderRadius: BorderRadius.circular(10),
                    ),
                  ),
              ],
            ));
      }),
    );
  }
}
